lib/pages/goal_page.dart

```diff
@@ -3,102 +3,529 @@ import 'package:linear/model/goal.dart';
 }
-
+ 
 class GoalPageState extends State<GoalPage> {
-  List<dynamic> _goals = [];
+ 
+    List<dynamic> _allGoals= [];
+    List<dynamic> _goals = [];
+    String sortingBy = "Newest";
+    String filteringBy= "Open";
+    bool isSorting = true;
 
-  bool _isLoading = true;
-  bool _isErrorFetchingGoals = false;
+    TextEditingController searchInput = TextEditingController();
 
-  @override
-  void initState() {
-    super.initState();
-    fetchGoals();
-  }
+  List<dynamic> _searchResults = [];
+  bool _hasPerformedSearch = false;
+   bool _isSearching = false;
+ 
+ bool _isLoading = true;
+ bool _isErrorFetchingGoals = false;
+ 
+ @override
+ void initState() {
+   super.initState();
+   fetchGoals();
+ }
+ 
+ fetchGoals() {
+   final Future<Map<String, dynamic>> successfulMessage = api.getGoalsForGoalPage(context);
+   successfulMessage.then((response) {
+     if (response['status'] == true) {
+       setState(() {
+         _allGoals = response['goals'];
+         _isLoading = true;
+         _isErrorFetchingGoals = false;
+       });
+       setState(() {
+         _goals = _allGoals.toList();
+       });
+     getOpen(_goals);
+     isSorting = true;
+     sortingBy = "Newest";
+     filteringBy= "Open";
+     setState(() {
+         _isLoading = false;
+         _isErrorFetchingGoals = false;
+       });
+ 
+     } else {
+       setState(() {
+          _allGoals = [];
+         _isLoading = false;
+         _isErrorFetchingGoals = true;
+       });
+       showDialog(
+            context: context,
+            builder: (context) {
+              return AlertDialog(
+                title: const Text("Error!"),
+                content: const Text(
+                    "An error occured while attempting to get the goals"),
+                actions: [
+                  TextButton(
+                      onPressed: () {
+                        Navigator.pop(context);
+                      },
+                      child: const Text("Ok"))
+                ],
+              );
+            });
+     }
+   });
+ }
 
-  fetchGoals() {
-    final Future<Map<String, dynamic>> successfulMessage = api.getGoalsForGoalPage(context);
-    successfulMessage.then((response) {
-      if (response['status'] == true) {
-        setState(() {
-          _goals = response['goals'];
-          _isLoading = false;
-          _isErrorFetchingGoals = false;
-        });
-      } else {
-        setState(() {
-          _goals = [];
-          _isLoading = false;
-          _isErrorFetchingGoals = true;
-        });
-      }
-    });
-  }
+ getOpen(List getGoals){
+  
+   setState(() {
+    _goals = _allGoals.toList();
+  });
+   Goal checkGoal = Goal.fromJson(_goals[0]);
+      for(int i = 0; i < _goals.length; i++){
+       checkGoal = Goal.fromJson(_goals[i]);
+       if((checkGoal.isFinished) == true){
+            setState(() {
+             _goals.removeAt(i);
+            i--;
+           });
+       }
+     }
```

Comment:
1. This assumes that I'll always have at least one goal? What I have no goals?
2. Instead of assigning the entire list then removing the ones that don't fit the criteria, add the ones that do into a local list and set the `_goals` to this local list.

Same comments apply to all of these "filter functions",